### ISSUE TITLE
bst1: buildstream is not compatible with python 3.11

### ISSUE THREAD

**Issue by jjardon** [apache/buildstream#1758]
See https://github.com/apache/buildstream/pull/1757

This is a problem as, for example,  the Fedora 37 default installation only includes python 3.11

Issue for master (buildstream2): #1802

**Comment by abderrahim** [comment]
It seems that ruamel.yaml.clib is the problem here, it doesn't provide a 3.11 wheel and fails to build with 3.11. But since it's an optional dependency, we can probably skip it for python 3.11 for now.

Sadly, it's not optional for bst2 so we'll have to wait for bst2.

**Comment by mcatanzaro** [comment]
I see the following error when running `bst build --track-all`:

```
[00:00:00][4043f233][build:freedesktop-sdk.bst:bootstrap/build/base-sdk/filtered.bst] BUG     Build

    An unhandled exception occured:
    
    Traceback (most recent call last):
      File "/usr/lib/python3.11/site-packages/buildstream/_scheduler/jobs/job.py", line 384, in _child_action
        result = self.child_process()
                 ^^^^^^^^^^^^^^^^^^^^
      File "/usr/lib/python3.11/site-packages/buildstream/_scheduler/jobs/elementjob.py", line 94, in child_process
        return self._action_cb(self._element)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/lib/python3.11/site-packages/buildstream/_scheduler/queues/buildqueue.py", line 35, in process
        return element._assemble()
               ^^^^^^^^^^^^^^^^^^^
      File "/usr/lib/python3.11/site-packages/buildstream/element.py", line 1534, in _assemble
        collect = self.assemble(sandbox)
                  ^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/lib/python3.11/site-packages/buildstream/plugins/elements/filter.py", line 108, in assemble
        dep.stage_artifact(sandbox, include=self.include,
      File "/usr/lib/python3.11/site-packages/buildstream/element.py", line 641, in stage_artifact
        files = list(self.__compute_splits(include, exclude, orphans))
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/lib/python3.11/site-packages/buildstream/element.py", line 2329, in __compute_splits
        self.__init_splits()
      File "/usr/lib/python3.11/site-packages/buildstream/element.py", line 2313, in __init_splits
        self.__splits = {
                        ^
      File "/usr/lib/python3.11/site-packages/buildstream/element.py", line 2314, in <dictcomp>
        domain: re.compile('^(?:' + '|'.join([utils._glob2re(r) for r in rules]) + ')$')
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/lib64/python3.11/re/__init__.py", line 227, in compile
        return _compile(pattern, flags)
               ^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/lib64/python3.11/re/__init__.py", line 294, in _compile
        p = _compiler.compile(pattern, flags)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/lib64/python3.11/re/_compiler.py", line 743, in compile
        p = _parser.parse(p, flags)
            ^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/lib64/python3.11/re/_parser.py", line 980, in parse
        p = _parse_sub(source, state, flags & SRE_FLAG_VERBOSE, 0)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/lib64/python3.11/re/_parser.py", line 455, in _parse_sub
        itemsappend(_parse(source, state, verbose, nested + 1,
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/lib64/python3.11/re/_parser.py", line 863, in _parse
        p = _parse_sub(source, state, sub_verbose, nested + 1)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/lib64/python3.11/re/_parser.py", line 455, in _parse_sub
        itemsappend(_parse(source, state, verbose, nested + 1,
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "/usr/lib64/python3.11/re/_parser.py", line 841, in _parse
        raise source.error('global flags not at the start '
   re.error: global flags not at the start of the expression at position 23
```

The problem is a broken regex here:

```
      File "/usr/lib/python3.11/site-packages/buildstream/element.py", line 2314, in <dictcomp>
        domain: re.compile('^(?:' + '|'.join([utils._glob2re(r) for r in rules]) + ')$')
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by mcatanzaro** [comment]
@nanonyme found https://github.com/apache/buildstream/commit/e8055a56d265d558248557782fa3ed0e6419caa4 which went into bst2. Probably need to backport that?

**Comment by nanonyme** [comment]
The CI result is bogus. pip installation of bst1 (or rather its dependency) ruamel.yaml.clib fails on Fedora 37. But this is an optional dependency which Fedora does not normally ship. So bst1 should be installable from distro packages.

**Comment by nanonyme** [comment]
It's possible https://sourceforge.net/p/ruamel-yaml-clib/code/ci/default/tree/_ruamel_yaml.c simple needs regenerating with newer cython.

**Comment by gtristan** [comment]
> It's possible https://sourceforge.net/p/ruamel-yaml-clib/code/ci/default/tree/_ruamel_yaml.c simple needs regenerating with newer cython.

Indeed revisioning cython generated code (if indeed that’s what this is) is not feasible, especially given that cython accesses internal/unstable apis from cpython.

**Comment by nanonyme** [comment]
@gtristan history says it was regenerated with cython

**Comment by nanonyme** [comment]
The problem is potentially now solved upstream. We should regenerate dependencies again.

**Comment by ghost** [comment]
Tried that, got these dependencies:

```
click==8.1.3
grpcio==1.50.0
Jinja2==3.1.2
pluginbase==1.0.1
protobuf==4.21.9
psutil==5.9.3
ruamel.yaml==0.17.21
ruamel.yaml.clib==0.2.7
setuptools==59.6.0
pyroaring==0.3.3
ujson==5.5.0
python-dateutil==2.8.2
## The following requirements were added by pip freeze:
MarkupSafe==2.1.1
six==1.16.0
```

[ruamel.yaml.clib 0.2.7](https://pypi.org/project/ruamel.yaml.clib/) fixes the cython issue for ruamel.yaml.

pyroaring now fails with the same issue. There is some discussion about python 3.11 support just starting: https://github.com/Ezibenroc/PyRoaringBitMap/issues/78

**Comment by nanonyme** [comment]
As said in IRC, I suggest submitting to pyroaring PR where relevant file is regenerated with newer Cython (probably same version as was used to fix ruamel.yaml.clib).

**Comment by nanonyme** [comment]
@ssssam also please create a separate issue about bst2 Python 3.11 support. This is about bst1. They have different requirements. Currently bst1 supports Python 3.11 whereas bst2 *does not* (because pyroaring does not support Python 3.11).

**Comment by nanonyme** [comment]
Neither is tested on Python 3.11 though yet.

**Comment by germinolegrand** [comment]
For people like me with bst1 in Fedora 37 and python 3.11 ; i quick fixed buildstream/element.py with the following:

buildstream/element.py:2370
```python
    def __file_is_whitelisted(self, path):
        # Considered storing the whitelist regex for re-use, but public data
        # can be altered mid-build.
        # Public data is not guaranteed to stay the same for the duration of
        # the build, but I can think of no reason to change it mid-build.
        # If this ever changes, things will go wrong unexpectedly.
        if not self.__whitelist_regex:
            bstdata = self.get_public_data('bst')
            whitelist = _yaml.node_get(bstdata, list, 'overlap-whitelist', default_value=[])
            whitelist_expressions = [
                utils._glob2re(
                    self.__variables.subst(
                        exp.strip(),
                        _yaml.node_get_provenance(bstdata, key='overlap-whitelist', indices=[index])
                    )
                )[5:] # FIX Skip flags "(?ms)" to avoid "global flags not at the start of the expression at position 4"
                for index, exp in enumerate(whitelist)
            ]
            expression = ('^(?:' + '|'.join(whitelist_expressions) + ')$')
            self.__whitelist_regex = re.compile(expression, flags=re.M|re.S) # FIX Add global flags
        return self.__whitelist_regex.match(path) or self.__whitelist_regex.match(os.path.join(os.sep, path))
```

This allowed me to open my dev_shell successfully.

**Comment by nanonyme** [comment]
Do you mean the bug is still not fixed? There are technical issues running CI against Python 3.11.  If you have a fix to code that doesn't result in regressions, pull requests welcome

**Comment by germinolegrand** [comment]
I dare not imply that. I did not test your PR because i'm a buildstream user only and i don't know how to build buildstream ; That is why i call that a "quick fix". I just wanted my installation (recently upgraded to Fedora 37) to work without waiting for your PR to land in a release available to me...

Although, just by reading the code, i'm not sure it would be sufficient to upgrade yaml ; because buildstream/element.py does something that is deprecated since Python 3.6 (and just turned into error in 3.11) : concatenate multiple regex expressions that have global flags like (?ms). My quick fix is a workaround, albeit a dirty one, not sure it is ok to put such a code in a production code base.

Would be better either to ask yaml not to generate the flags, or ensure the newer yaml version does not generate them anymore, or to make sure it ALWAYS generates exactly "(?ms)" at the beginning.

**Comment by nanonyme** [comment]
The backport I did should be in Fedora 37 already. Open PR is just about fixing CI to test against newer platform.

**Comment by germinolegrand** [comment]
I am currently using bst 1.6.8.
I check if any fedora updates are available and will keep you posted.

**Comment by nanonyme** [comment]
That is indeed the latest version.

**Comment by nanonyme** [comment]
@mcatanzaro was your original problem fixed?

**Comment by germinolegrand** [comment]
Reading the element.py on master it seems this exact problem is solved (not related to yaml at all) in bst2.

```python
    # _file_is_whitelisted()
    #
    # Checks if a file is whitelisted in the overlap whitelist
    #
    # This is only internal (one underscore) and not locally private
    # because it needs to be proxied through ElementProxy.
    #
    # Args:
    #    path (str): The path to check
    #
    # Returns:
    #    (bool): True of the specified `path` is whitelisted
    #
    def _file_is_whitelisted(self, path):
        # Considered storing the whitelist regex for re-use, but public data
        # can be altered mid-build.
        # Public data is not guaranteed to stay the same for the duration of
        # the build, but I can think of no reason to change it mid-build.
        # If this ever changes, things will go wrong unexpectedly.
        if not self.__whitelist_regex:
            bstdata = self.get_public_data("bst")
            whitelist = bstdata.get_sequence("overlap-whitelist", default=[])
            whitelist_expressions = [utils._glob2re(self.__variables.subst(node)) for node in whitelist]
            expression = "^(?:" + "|".join(whitelist_expressions) + ")$"
            self.__whitelist_regex = re.compile(expression, re.MULTILINE | re.DOTALL)
        return self.__whitelist_regex.match(os.path.join(os.sep, path))
```

Git blame shows it was this commit that fixed it in utils._glib2re which was the culprit :

https://github.com/apache/buildstream/commit/782555a674de2ab6f1760f13c4a1b17400f64533#diff-ddfabb836a421fd2981aea75354b2c11dba5343df2543c56124a8393aa651cc5

This is a much better fix than mine.

**Comment by germinolegrand** [comment]
And also, you did fix utils._glob2re in bst1 here 4755665c3665bbaba8794b35acf69f2d7b9a8ac3

I don't understand why it did not happened to be in my 1.6.8 although it is commited on the bst-1 branch and in the 1.6.8 tag history ??

**Comment by germinolegrand** [comment]
I read your commit https://github.com/apache/buildstream/commit/4755665c3665bbaba8794b35acf69f2d7b9a8ac3 badly. It actually introduces the bug ! (or at least it does not fix it : you should not put (?ms) at all like bst2 does, and add global flags on call sites)

**Comment by nanonyme** [comment]
We're pretty sure (?ms) was put into wrong place before that commit. Would you like to make a PR with the correct backport? You can just pip install the project from git to test your change locally.

**Comment by nanonyme** [comment]
@germinolegrand I ended up just doing blind backport. Can you try if this helps for you? https://github.com/apache/buildstream/pull/1800/files (`pip install .`)

**Comment by mcatanzaro** [comment]
> @mcatanzaro was your original problem fixed?

Yes, I don't need to use toolbx anymore.

**Comment by nanonyme** [comment]
Okay, so this is clearly a follow-up bug.

**Comment by germinolegrand** [comment]
@nanonyme I tried your PR code, it worked for me !

**Comment by mcatanzaro** [comment]
> Okay, so this is clearly a follow-up bug.

Unfortunately, today I'm hitting almost the original bug again, so apparently not. The entire backtrace is identical to the one that I posted above except for the final line, which is now `re.error: global flags not at the start of the expression at position 4`. The error was previously at position 23 but is now at position 4. Hopefully your #1800 will fix it.

**Comment by nanonyme** [comment]
@mcatanzaro can you try `pip install .` with that branch? CI is currently not available for Fedora 37 so these are kind of blind fixes.

**Comment by mcatanzaro** [comment]
No, sorry, messing with pip and custom buildstream installs requires time and effort. I can test Fedora packaged builds, though.

The history of problems with python could be avoided with better CI. buildstream should test upcoming python versions well before they are released, not after. I think testing with Fedora rawhide might be adequate to achieve this? You'll never get sufficient test coverage by using numbered releases unless your CI builds its own python.

**Comment by nanonyme** [comment]
Yeah, well, this has been emitting deprecation warnings for ages. It was not unannounced breakage.

**Comment by jjardon** [comment]
> @mcatanzaro can you try pip install . with that branch? CI is currently not available for Fedora 37 so these are kind of blind fixes.

@nanonyme can you maybe rebase your PR https://github.com/apache/buildstream/pull/1800 on top of https://github.com/apache/buildstream/pull/1757 so we know test pass (it will test in F37 and python3.11)

**Comment by nanonyme** [comment]
@jjardon no because your branch is missing all the Python 3.11 support work. You have invalid dependencies. When dependencies are updated so test harness can start, it hangs.

**Comment by nanonyme** [comment]
Well, okay does not hang but does not finish either. I have some more work in https://github.com/apache/buildstream/pull/1785 about it. Latest build is failing because BuildStream is relying on something that absolutely cannot be supported IOW running BuildStream main loop repeatedly in pytest.

**Comment by nanonyme** [comment]
I was trying to make BuildStream instead be launched as external executable and found out that the tests are relying on living in same process with BuildStream so they can extract failures from stack.

**Comment by nanonyme** [comment]
BuildStream 2 is broken in different ways IOW pyroaring does not yet support Python 3.11. We have absolutely no coverage on any BuildStream branch on modern Python versions.

**Comment by mcatanzaro** [comment]
Anything more needed here? Surely it's time to release a fix for this issue?

**Comment by gtristan** [comment]
> Anything more needed here? Surely it's time to release a fix for this issue?

No this is not trivial.

I came back from holiday Monday and spent a bunch of hours investigating this this week, but ran out of time due to other obligations, I’ll try to focus on this as soon as I can.

Unfortunately we use grpc, which is a steaming pile of google trash which breaks in mysterious ways multiple times a year, and won’t pass BuildStream tests at the latest grpc version even with Python 3.10, grpc for Python 3.11 needs to be the more recent broken version.

I don’t think overall it’s realistic to try using fresh python versions without letting the ecosystem harden for at least 6 months or so, but we do cook docker images, which should be helpful for those who really want bleeding edge distros on their hosts.

**Comment by nanonyme** [comment]
@gtristan I would *again* note it's not BuildStream+grpcio that break, it's the unit tests which are running BuildStream in same process repeatedly. This is not done by normal users. @mcatanzaro has already been testing BuildStream with latest fixes and could build projects with it.

**Comment by nanonyme** [comment]
The CI should probably be rewritten from scratch to run BuildStream commands in new processes.

**Comment by mcatanzaro** [comment]
> I don’t think overall it’s realistic to try using fresh python versions without letting the ecosystem harden for at least 6 months or so

It's super annoying for GNOME release team to be unable to use buildstream outside a special container environment. buildstream is supposed to _itself_ be the container environment....

**Comment by nanonyme** [comment]
I thought these were just regular Fedora containers with BuildStream and its deps distributed with them with supported versions.

**Comment by gtristan** [comment]
> > I don’t think overall it’s realistic to try using fresh python versions without letting the ecosystem harden for at least 6 months or so
> 
> It's super annoying for GNOME release team to be unable to use buildstream outside a special container environment. buildstream is supposed to _itself_ be the container environment....

I agree, but I don’t think anyone is mandating that release team members use latest fedora either (and I’m not sure that it’s suitable for fedora to wrap such a new python version into a “stable” release either).

**Comment by gtristan** [comment]
> @gtristan I would _again_ note it's not BuildStream+grpcio that break, it's the unit tests which are running BuildStream in same process repeatedly. This is not done by normal users. @mcatanzaro has already been testing BuildStream with latest fixes and could build projects with it.

Yes I’ve heard that, still will take a bit of fiddling to try that out, tests have to pass or we don’t know.

asides from that, this is just *one* of the plethora of ways grpc likes to constantly break its consumers, we’re constantly pinning it due to flagrant instability, as long as that doesn’t change we’re still gonna have this problem with the bleeding edge.

**Comment by nanonyme** [comment]
And yes, I agree we cannot claim Python 3.11 support or close this issue without fixing the tests. The question is whether to ship the regexp fixes which allow bst to work *at all* on Python 3.11 or not. Currently it doesn't work because our CI doesn't fail on deprecation warnings so we've been leaving more than a year worth of deprecations unfixed in this project.

**Comment by mcatanzaro** [comment]
> I agree, but I don’t think anyone is mandating that release team members use latest fedora either 

I spend half my time on the _next_ stable Fedora, so using the latest Fedora is slow for me. Using a previous stable release is out of the question. We need to develop the next one.

> (and I’m not sure that it’s suitable for fedora to wrap such a new python version into a “stable” release either).

Similarly, Fedora often ships the _next_ stable python. This is essential to ensure the quality of python releases; otherwise, if no distros ship until the stable release, how to ensure the stable release actually works with a large body of python software? Same reason we build spring Fedoras with prerelease GCC, to ensure the GCC release is good. python 3.12 is just around the corner now.

**Comment by abderrahim** [comment]
> Unfortunately we use grpc, which is a steaming pile of google trash which breaks in mysterious ways multiple times a year, and won’t pass BuildStream tests at the latest grpc version even with Python 3.10, grpc for Python 3.11 needs to be the more recent broken version.

Just floating the idea here: what do you think of porting to [grpclib](https://grpclib.readthedocs.io/en/latest/) instead of grpcio? It claims to be pure python so it should be less prone to breaking with newer python versions.

**Comment by nanonyme** [comment]
I'm suspicious about performance impact on bst1. With bst2 it might be less due to usage of buildbox-casd as proxy. But maybe it would for maintenance reasons still be worth it.

**Comment by nanonyme** [comment]
We also *claim to* support Python 3.6 for bst1 and grpclib requires Python 3.7 or newer.

**Comment by nanonyme** [comment]
Note that BuildStream is now broken with Fedora 37. Fedora 38 is going to be released soon which means that there will soon not be non-EOL Fedora that BuildStream supports. This should be a serious matter for BuildStream upstream if it regards BuildStream 1.6 maintained.

**Comment by mcatanzaro** [comment]
@gtristan can you please tag a 1.6 release? I believe that's all that's needed to end the emergency.

**Comment by mcatanzaro** [comment]
I'm at wits end here. Are there no maintainers willing to make a release?

**Comment by mcatanzaro** [comment]
I'm bamboozled as to what the hold-up is here and am no longer willing to wait. @gtristan if you don't release by the end of this week, I will create a fork of buildstream on gitlab.gnome.org with no source code changes and release there. Would be much easier if you would simply tag a release.

**Comment by nanonyme** [comment]
To be fair, that might not be so horrible outcome. BuildStream 1 is not an Apache project.

**Comment by nanonyme** [comment]
You may want to only grab bst-1 branch if you end up doing that.

**Comment by mcatanzaro** [comment]
Having an extra project is going to be confusing, so it would be more ideal to release here IMO. But any release is better than leaving the current release broken.

> You may want to only grab bst-1 branch if you end up doing that.

That's the plan.

**Comment by nanonyme** [comment]
Sure. I'm not a project member, I'm not telling whether or not to do that. I would just prefer keeping things clean if that is done.

**Comment by jjardon** [comment]
> I'm bamboozled as to what the hold-up is here and am no longer willing to wait. @gtristan if you don't release by the end of this week, I will create a fork of buildstream on gitlab.gnome.org with no source code changes and release there. Would be much easier if you would simply tag a release.

@mcatanzaro I can sympathise with you and I know this is not ideal (remember I have opened this issue!), but I'm aware tristan has been extremely busy at work this last weeks. Seems the situation is better now so let's have a bit of patience please

**Comment by nanonyme** [comment]
@jjardon regardless of workload for @gtristan it is a *big* problem that the fundamental component for building freedesktop-sdk and gnome runtimes seems to have this low bus factor. It was just presented in FOSDEM that flatpak is more or less winning the desktop application race. Now is not the time to slow down, now is time to ramp up speed and make even better software platforms.

**Comment by jjardon** [comment]
> @jjardon regardless of workload for @gtristan it is a _big_ problem that the fundamental component for building freedesktop-sdk and gnome runtimes seems to have this low bus factor. It was just presented in FOSDEM that flatpak is more or less winning the desktop application race. Now is not the time to slow down, now is time to ramp up speed and make even better software platforms.

I do not disagree

**Comment by nanonyme** [comment]
Should this now be closed as @gtristan got CI to work with bst1 and there is a release tag with the fixes?

**Comment by mcatanzaro** [comment]
I think so. Thanks @gtristan !

**Comment by jjardon** [comment]
Fixed with https://github.com/apache/buildstream/pull/1827

**Comment by gtristan** [comment]
This is fixed by #1827.
